Expose OpenSSL interfaces for base64 coding.

adamreichold opened the pull request:
This exposes the `EVP_EncodeBlock` and `EVP_DecodeBlock` API of OpenSSL which can be used for base64 coding.

It seems reasonable to include those as they have no interactions with and hence unintended side-effects on the rest of the library and it would allow crates already depending on this crate to avoid additional dependencies for base64 coding.

sfackler commented on openssl/src/base64.rs:
```diff
@@ -0,0 +1,110 @@
+    }
+}
+
+fn encoded_len(src_len: c_int) -> Option<usize> {
```
Why does this take a c_int and return a usize?

adamreichold replied:
I need the return value to `reserve`/`resize` the `out` buffer, so I need an `usize`. But the computation must succeed using a `c_int` for the OpenSSL calls to return valid counts otherwise I am not convinced I can rely on those counts for `set_len`/`truncate`.

In an earlier version, I used `c_int -> Option<c_int>` and had `encoded_len(src_len).unwrap().try_into().unwrap()` on the outside which seemed cumbersome but indeed less surprising. I can certainly changed this back if you prefer or I can add comment explaining the odd types better?

adamreichold replied:
But I just noticed that the CI builds failed as the use Rust 1.33 which I guess is this libraries current minimum supported Rust version so I guess I have to remove the `TryInto` usage in any case?

adamreichold replied:
Pushed a version that builds using Rust 1.33 and added a comment explaining the odd types.

Would you prefer the `usize_to_c_int` and `c_int_to_usize` in the top-level module or replaced by plain asserts as used in the other modules?

sfackler replied:
It seems best to mirror the rest of the codebase.

adamreichold replied:
Pushed a version that hopefully better mirrors the style of the rest of the code base.

sfackler commented on openssl/src/base64.rs:
```diff
@@ -0,0 +1,129 @@
+    // `EVP_EncodeBlock` will only write to not read from `out`.
+    unsafe {
+        let out_len = ffi::EVP_EncodeBlock(out.as_mut_ptr(), src.as_ptr(), src_len);
+        out.set_len(c_int_to_usize(out_len).unwrap());
```
This can just be `out_len as usize`.

adamreichold replied:
Indeed. Fixed this and the checked cast in `{de,en}coded_len` that was actually unnecessary as well as we always call this with non-negative values and the arithmetic is checked.

adamreichold commented:
During the last rebase, I also noticed that I actually forgot the `#[test]` attribute on two of tests which I now added but which also showed that I forgot to strip the zero padding from the decoded output which I added as well.

sfackler commented on openssl/src/base64.rs:
```diff
@@ -0,0 +1,125 @@
+        out.set_len(out_len as usize);
+    }
+
+    if src.ends_with("=") {
```
What is this doing?

sfackler replied:
Oh right, null padding.
